gno.land/pkg/integration/pkgloader.go

```diff
@@ -0,0 +1,168 @@
+package integration
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/gnolang/gno/gno.land/pkg/gnoland"
+	"github.com/gnolang/gno/gno.land/pkg/sdk/vm"
+	"github.com/gnolang/gno/gnovm/pkg/gnolang"
+	"github.com/gnolang/gno/gnovm/pkg/gnomod"
+	"github.com/gnolang/gno/gnovm/pkg/packages"
+	bft "github.com/gnolang/gno/tm2/pkg/bft/types"
+	"github.com/gnolang/gno/tm2/pkg/std"
+)
+
+type PkgsLoader struct {
+	pkgs    []gnomod.Pkg
+	visited map[string]struct{}
+
+	// list of occurrences to patchs with the given value
+	// XXX: find a better way
+	patchs map[string]string
+}
+
+func NewPkgsLoader() *PkgsLoader {
+	return &PkgsLoader{
+		pkgs:    make([]gnomod.Pkg, 0),
+		visited: make(map[string]struct{}),
+		patchs:  make(map[string]string),
+	}
+}
+
+func (pl *PkgsLoader) List() gnomod.PkgList {
+	return pl.pkgs
+}
+
+func (pl *PkgsLoader) SetPatch(replace, with string) {
+	pl.patchs[replace] = with
+}
+
+func (pl *PkgsLoader) LoadPackages(creator bft.Address, fee std.Fee, deposit std.Coins) ([]gnoland.TxWithMetadata, error) {
+	pkgslist, err := pl.List().Sort() // sorts packages by their dependencies.
+	if err != nil {
+		return nil, fmt.Errorf("unable to sort packages: %w", err)
+	}
+
+	txs := make([]gnoland.TxWithMetadata, len(pkgslist))
+	for i, pkg := range pkgslist {
+		tx, err := gnoland.LoadPackage(pkg, creator, fee, deposit)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load pkg %q: %w", pkg.Name, err)
+		}
+
+		// If any replace value is specified, apply them
+		if len(pl.patchs) > 0 {
+			for _, msg := range tx.Msgs {
+				addpkg, ok := msg.(vm.MsgAddPackage)
+				if !ok {
+					continue
+				}
+
+				if addpkg.Package == nil {
+					continue
+				}
+
+				for _, file := range addpkg.Package.Files {
+					for replace, with := range pl.patchs {
+						file.Body = strings.ReplaceAll(file.Body, replace, with)
+					}
+				}
+			}
+		}
+
+		txs[i] = gnoland.TxWithMetadata{
+			Tx: tx,
+		}
+	}
+
+	return txs, nil
+}
+
+func (pl *PkgsLoader) LoadAllPackagesFromDir(path string) error {
+	// list all packages from target path
+	pkgslist, err := gnomod.ListPkgs(path)
+	if err != nil {
+		return fmt.Errorf("listing gno packages: %w", err)
+	}
+
+	for _, pkg := range pkgslist {
+		if !pl.exist(pkg) {
+			pl.add(pkg)
+		}
+	}
+
+	return nil
+}
+
+func (pl *PkgsLoader) LoadPackage(modroot string, path, name string) error {
```

Comment:
I'm not even gonna ask if we have unified logic for this somewhere :)

Reply:
Yes, unfortunately... But there is an ongoing effort made by @n0izn0iz to unify this logic.